mantidproject/mantid
# Settings widget overhaul

### Description of work
Added models to each of the settings widget tabs to handle changes to configs.

#### Summary of work
Before, changes in the settings widget would immediately be written to file.
Now, changes are stored in the models and only written to file when either of the new 'apply' or 'okay' buttons are pressed.
If you attempt to close the settings widget without applying your changes, you'll be asked to confirm you want to close.

Fixes #38293

#### Further detail of work
Also did some minor moving of buttons on the setting widget (load and save to file have been moved to the left).
I've also stopped the combo and spin boxes from responding to mouse wheel events (this was very annoying when scrolling down the settings you would change a bunch of them by accident)

There are further improvements that could be made:
 - More strictly following mvp by handling connections in the view (and probably other things)
 - Putting an `*` or some kind of indication on the settings you've changed in the gui.

### To test:

I think its best to just test as many logical paths as you can think of i.e:
- Change some settings, apply and close (check they've had an effect)
- Change settings and click close to check you're warned (also check 'yes' and 'no' do what's expected
- Change a setting then set it back to its original value, you shouldn't be warned about closing the widget
- Load and save some settings to file (I haven't touched this, but good to check)
- Anything else you can think of.



---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
